daemon/image_delete.go

```diff
@@ -160,21 +175,64 @@ func (daemon *Daemon) canDeleteImage(imgID string, force bool) error {
 			return err
 		}
 
+		direct := true
 		if err := daemon.graph.WalkHistory(parent, func(p image.Image) error {
 			if imgID == p.ID {
-				if container.IsRunning() {
-					if force {
-						return fmt.Errorf("Conflict, cannot force delete %s because the running container %s is using it, stop it and retry", stringid.TruncateID(imgID), stringid.TruncateID(container.ID))
-					}
-					return fmt.Errorf("Conflict, cannot delete %s because the running container %s is using it, stop it and use -f to force", stringid.TruncateID(imgID), stringid.TruncateID(container.ID))
-				} else if !force {
-					return fmt.Errorf("Conflict, cannot delete %s because the container %s is using it, use -f to force", stringid.TruncateID(imgID), stringid.TruncateID(container.ID))
+				if direct {
+					directlyUsedBy = append(directlyUsedBy, container.ID)
+				} else {
+					indirectlyUsedBy = append(indirectlyUsedBy, container.ID)
 				}
 			}
+			direct = false
 			return nil
 		}); err != nil {
 			return err
 		}
 	}
-	return nil
+
+	if len(directlyUsedBy) == 0 && len(indirectlyUsedBy) == 0 {
+		// The image can be deleted because no containers are using the image.
+		return nil
+	} else if len(directlyUsedBy) != 0 {
+		if len(indirectlyUsedBy) != 0 {
+			// The image is a history layer of another image.
+			// At the same, there are some containers using it.
+			// If the -f flag is provided we can try to untag it.
+			// But the data of the image should not be deleted irrespective of the -f flag.
+			if force {
+				return errUntagable
+			}
+			return fmt.Errorf("Conflict, cannot delete %s because the container %s is using it. But you can use -f to untag it", stringid.TruncateID(imgID), stringid.TruncateID(directlyUsedBy[0]))
+		}
+
+		// The image is directly used by some containers.
+		// It depends on the state of the containers and -f flag.
+		for _, cid := range directlyUsedBy {
+			container, err := daemon.Get(cid)
+			if err != nil {
+				if truncindex.IsNotExist(err) {
+					// The container might had been deleted after the previous call to daemon.List.
+					continue
+				}
+				return err
+			}
+
+			if container.IsRunning() {
```

Comment:
Another case where we have unnecessary indentation.

```
if !container.IsRunning() {
    if !force {
        return fmt.Errorf("Conflict, cannot delete %s because the container %s is using it, use -f to force", stringid.TruncateID(imgID), stringid.TruncateID(cid))
    }

    continue
}
```

Double check the logic above, but it will make this loop more digestible.